### PR TITLE
Fixed #29692 -- Fixed removing ordering parts for multiline RawSQL expressions.

### PR DESCRIPTION
**Pull request by cansarigol**
Ticket: [#29692](https://code.djangoproject.com/ticket/29692)

### CODE REVIEW

File: tests/queries/tests.py [hunk 1]
```diff
@@ -1629,6 +1630,21 @@ def test_ordering(self):
             [2, 1, 3]
         )
 
+        # newlines control for raw sql parse of compiler.get_order_by
```
Comment (charettes) [inline tests/queries/tests.py]: Please use a more appropriate location for the test. `tests/expressions/tests.py` seems like a better candidate.

Comment (cansarigol) [reply tests/queries/tests.py]: I've moved the test into `tests/expressions/tests.py`, thanks.

File: django/db/models/sql/compiler.py [hunk 2]
```diff
@@ -356,7 +356,7 @@ def get_order_by(self):
             # not taken into account so we strip it. When this entire method
             # is refactored into expressions, then we can check each part as we
             # generate it.
-            without_ordering = self.ordering_parts.search(sql).group(1)
+            without_ordering = self.ordering_parts.search("".join(sql.splitlines())).group(1)
```
Comment (timgraham) [inline django/db/models/sql/compiler.py]: Use single quotes (unless a string contains a single quote) as described in [Python coding style](https://docs.djangoproject.com/en/dev/internals/contributing/writing-code/coding-style/#python-style).

It seems a code comment may also be useful.

File: tests/expressions/tests.py [hunk 3]
```diff
@@ -380,6 +380,21 @@ def test_order_by_exists(self):
         )
         self.assertSequenceEqual(mustermanns_by_seniority, [self.max, mary])
 
+    def test_order_by_rawsql(self):
+        self.assertQuerysetEqual(
```
Comment (timgraham) [inline tests/expressions/tests.py]: `assertSequenceEqual` is preferred.

**Comment by cansarigol** [comment]
I didn't think that the test would fail hence I checked just SQLite. it's weird that the interpretation of null value in order clause of Postgresql as well as Oracle.

```
CREATE TEMP TABLE temp_order_table (id int);
insert into temp_order_table(id) values(1);
insert into temp_order_table(id) values(null);
select * from temp_order_table order by id desc
>> id
   ----
   NULL
   1
```
So I've changed the `case ... else NULL end` phrase.

**Comment by cansarigol** [comment]
> Does the test only cover the first place that changed? 
Is this test missing or do we need a lot of tests to cover all cases of raw sql that should be?

**Comment by carltongibson** [comment]
Hi @cansarigol. I think Tim's point here is that the branch in `get_extra_select()` is not hit by your test case. (It needs a distinct query with a non-empty order_by...)

**Comment by cansarigol** [comment]
hi @carltongibson thanks, I added test for `get_extra_select()` as well.

**Comment by felixxm** [comment]
To be honest I don't like this solution :man_shrugging: I'm not comfortable with adding another workaround just for a single edge-case with `RawSQL`. IMO we should rather sanitize `sql` passed to `RawSQL`, i.e. replace `\n` and `\r` with a single space and multiple spaces with one, e.g.
```diff
diff --git a/django/db/models/expressions.py b/django/db/models/expressions.py
index 16df317631..5a95a66f2e 100644
--- a/django/db/models/expressions.py
+++ b/django/db/models/expressions.py
@@ -678,6 +678,7 @@ class RawSQL(Expression):
         if output_field is None:
             output_field = fields.Field()
         self.sql, self.params = sql, params
+        self.sql = ' '.join(self.sql.split())
         super().__init__(output_field=output_field)
 
     def __repr__(self):
```

It would also fix issue with difference number of spaces:
 - `(case when num_employees > 1000 then num_chairs else 0 end) DESC`
 - `(case when num_employees < 1000 then num_chairs else 0 end) ASC`

**Comment by cansarigol** [comment]
Would we do something to keep query legibility and prevent to perform the split every time?  just a suggestion, we can use a function (`django.utils`) where needed.

**Comment by felixxm** [comment]
I'm confused, which function from `django.utils` do you mean? IMO it's better to remove newlines directly in `RawSQL.__init__()` e.g. `'  '.join(sql.sql.splitlines())` rather then adding `RawSQL`-workarounds in multiple places.

**Comment by cansarigol** [comment]
if we change `RawSQL` directly
- the query string will be illegible.
- even it is not necessary, every time split will be performed.
Because of these, I thought that we can override the `search` function as our needs.

**Pull request merged**